src/Support/ModularServiceProvider.php

```diff
@@ -45,31 +45,40 @@ public function __construct($app)
 	
 	public function register(): void
 	{
-		$this->mergeConfigFrom("{$this->base_dir}/config.php", 'app-modules');
+		$this->mergeConfigFrom("{$this->base_dir}/config/app-modules.php", 'app-modules');
 		
-		$this->app->singleton(ModuleRegistry::class, function() {
+		$this->app->singleton(ModuleRegistry::class, function(Application $app) {
 			return new ModuleRegistry(
 				$this->getModulesBasePath(),
-				$this->app->bootstrapPath('cache/modules.php')
+				$app->make(AutodiscoveryHelper::class),
 			);
 		});
 		
-		$this->app->singleton(AutoDiscoveryHelper::class);
+		$this->app->singleton(FinderFactory::class, function() {
+			return new FinderFactory($this->getModulesBasePath());
+		});
 		
-		$this->app->singleton(MakeMigration::class, function($app) {
+		$this->app->singleton(AutodiscoveryHelper::class, function(Application $app) {
+			return new AutodiscoveryHelper(
+				$app->make(FinderFactory::class),
+				$app->make(Filesystem::class),
+				$this->app->bootstrapPath('cache/app-modules.php')
+			);
+		});
+		
+		$this->app->singleton(MakeMigration::class, function(Application $app) {
 			return new MigrateMakeCommand($app['migration.creator'], $app['composer']);
 		});
 		
 		$this->registerEloquentFactories();
 		
-		// Set up lazy registrations for things that only need to run if we're using
-		// that functionality (e.g. we only need to look for and register migrations
-		// if we're running the migrator)
-		$this->registerLazily(Migrator::class, [$this, 'registerMigrations']);
-		$this->registerLazily(Gate::class, [$this, 'registerPolicies']);
+		$this->app->resolving(Migrator::class, fn(Migrator $migrator) => $this->autodiscover()->migrations($migrator));
+		$this->app->resolving(Gate::class, fn(Gate $gate) => $this->autodiscover()->policies($gate));
 		
-		// Look for and register all our commands in the CLI context
-		Artisan::starting(Closure::fromCallable([$this, 'onArtisanStart']));
+		Artisan::starting(function(Artisan $artisan) {
+			$this->autodiscover()->commands($artisan);
+			$this->registerNamespacesInTinker();
+		});
```

Comment:
I would recommend that the `modules:cache` command be added to the `optimize` command:


```suggestion
		Artisan::starting(function(Artisan $artisan) {
			$this->autodiscover()->commands($artisan);
			$this->registerNamespacesInTinker();
		});
		
		$this->optimizes('modules:cache', 'modules:clear', 'modules');
```